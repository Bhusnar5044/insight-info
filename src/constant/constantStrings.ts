export const emptyStateMessages = {
    application: 'No applications available.',
    resources: 'No Resources available.',
    noResultFound: 'No result found',
};
