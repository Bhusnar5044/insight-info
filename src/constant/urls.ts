const baseURL = 'https://engineering-task.elancoapps.com/api';

export const urls = {
    login: 'https://api.escuelajs.co/api/v1/auth/login',
    profile: 'https://api.escuelajs.co/api/v1/auth/profile',
    raw: `${baseURL}/raw`,
    applications: `${baseURL}/applications`,
    resources: `${baseURL}/resources`,
};
