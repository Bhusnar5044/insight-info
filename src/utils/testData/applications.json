["dummyTitle", "dummyTitle2", "dummyTitle3"]
