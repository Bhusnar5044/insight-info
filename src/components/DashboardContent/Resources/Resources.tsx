import Loader from '@components/Loader';
import { SearchBox } from '@components/common';
import { urls } from '@constant/urls';
import { fetch } from '@utils';
import { FC, memo, useCallback, useEffect, useState } from 'react';
import { StyledCard, StyledCardContainer } from '../styled';

export const Resources: FC = memo(() => {
    const [isLoading, setIsLoading] = useState(false);
    const [resources, setResources] = useState([]);
    const [filteredResources, setFilteredResources] = useState(resources);

    const getResources = useCallback(async () => {
        setIsLoading(true);
        const { response, error } = await fetch({ url: urls.resources });
        if (response?.data) {
            setResources(response?.data);
            setFilteredResources(response?.data);
        }
        if (error) console.log(error);

        setIsLoading(false);
    }, []);

    const onSearch = (value: string) => {
        setFilteredResources(resources.filter((item: string) => item.toLowerCase().includes(value.toLowerCase())));
    };

    useEffect(() => {
        !resources.length && getResources();
    }, [getResources, resources.length]);

    return (
        <StyledCardContainer flexDirection="column" gap="1rem">
            <SearchBox onSearch={onSearch} placeholder="Search" />
            {isLoading && <Loader />}
            {!isLoading &&
                filteredResources.length &&
                filteredResources.map((resource, index) => (
                    <StyledCard key={`${resource}-${index}`} to={`/resources/${resource}`}>
                        {resource}
                    </StyledCard>
                ))}
        </StyledCardContainer>
    );
});
