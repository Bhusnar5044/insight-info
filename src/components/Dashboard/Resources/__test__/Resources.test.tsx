import { emptyStateMessages } from '@constant/constantStrings';
import { urls } from '@constant/urls';
import { fireEvent, mockAxios, renderWithRouter, screen, waitFor } from '@test-utils';
import * as mockData from '@utils/testData/applications.json';
import { Resources } from '../Resources';

const renderer = () => renderWithRouter(<Resources />);
describe('Resources', () => {
    beforeEach(() => {
        mockAxios.resetHistory();
        jest.clearAllMocks();
    });

    it('should render properly', () => {
        const { container } = renderer();
        expect(container).toMatchSnapshot();
    });

    it('should show no result available message on failure or in case response is empty', async () => {
        mockAxios.onGet(urls.applications).reply(200, []);
        renderer();
        expect(screen.getByTitle('loader')).toBeInTheDocument();
        await waitFor(() => {
            expect(screen.queryByTitle('loader')).not.toBeInTheDocument();
            expect(screen.getByText(emptyStateMessages.resources)).toBeInTheDocument();
        });
    });

    it('should fetch all the resources', async () => {
        mockAxios.onGet(urls.resources).reply(200, mockData);
        renderer();
        expect(screen.getByTitle('loader')).toBeInTheDocument();
        await waitFor(() => {
            expect(screen.queryByTitle('loader')).not.toBeInTheDocument();
            expect(screen.getByText('dummyTitle')).toBeInTheDocument();
        });
    });

    it('should show the result based on search', async () => {
        mockAxios.onGet(urls.applications).reply(200, mockData);
        renderer();
        expect(screen.getByTitle('loader')).toBeInTheDocument();
        await waitFor(() => {
            expect(screen.queryByTitle('loader')).not.toBeInTheDocument();
            expect(screen.getByText(/dummyTitle2/i)).toBeInTheDocument();
            const searchBox = screen.getByTitle('searchBox');
            fireEvent.change(searchBox, { target: { value: 'dummyTitle3' } });
            fireEvent.keyDown(searchBox, { key: 'Enter' });
            screen.debug();
            expect(screen.queryByTitle(/dummyTitle2/i)).not.toBeInTheDocument();
            expect(screen.getByText(/dummyTitle3/i)).toBeInTheDocument();
        });
    });
});
