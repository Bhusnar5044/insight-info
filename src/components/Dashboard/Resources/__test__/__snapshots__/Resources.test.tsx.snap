// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Resources should render properly 1`] = `
.c6 {
  width: 10rem;
  height: auto;
  margin: auto;
  animation: gOaCMp infinite 20s linear;
}

.c0 {
  display: flex;
  flex-direction: column;
  gap: 1rem;
}

.c5 {
  display: flex;
  gap: 0.5rem;
}

.c4 {
  -webkit-appearance: none;
  appearance: none;
  width: 100%;
  border: 0;
  font-family: inherit;
  font-size: 16px;
  font-weight: 400;
  background: #ffff;
  box-shadow: inset 0 -1px 0 rgba(#212529, 0.3);
  color: #212529;
  transition: all 0.15s ease;
  outline: none;
}

.c2 {
  position: relative;
  border-radius: 3px;
  padding: 0.5rem 1rem;
  display: flex;
  align-items: center;
  overflow: hidden;
  width: 23rem;
  background-color: #ffff;
}

.c2 .focus-bg {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background: rgba(#6c757d, 0.05);
  z-index: -1;
  transform: scaleX(0);
  transform-origin: left;
}

.c2 .c3 {
  width: 20rem;
}

.c2 img {
  cursor: pointer;
  width: 1.2rem;
  height: 1.2rem;
}

.c1 {
  margin: 2rem 0;
}

<div>
  <div
    class="c0 c1"
    flexdirection="column"
    gap="1rem"
  >
    <div
      class="c2"
      for="searchInput"
    >
      <input
        class="c3 c4"
        id="searchInput"
        placeholder="Search"
        title="searchBox"
        type="text"
        value=""
      />
      <div
        class="c5"
        gap="0.5rem"
      >
        <img
          alt="search-icon"
        />
      </div>
    </div>
    <img
      class="c6"
      title="loader"
    />
  </div>
</div>
`;
